### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_vrf_af.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ce_vrf_af.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```

TASK [Configure vpna, delete address family is ipv4] ***************************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_vrf_af.yml:55
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_vrf_af.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936 `" && echo ansible-tmp-1487054055.55-179097916783936="` echo $HOME/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmpOLQQxI TO /root/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936/ce_vrf_af.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936/ /root/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936/ce_vrf_af.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936/ce_vrf_af.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487054055.55-179097916783936/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "evpn_target_type": [], 
        "evpn_target_value": [], 
        "route_distinguisher": [], 
        "vpn_target_type": [], 
        "vpn_target_value": [], 
        "vrf": "vpna", 
        "vrf_aftype": []
    }, 
    "existing": {
        "evpn_target_type": [], 
        "evpn_target_value": [], 
        "route_distinguisher": [
            null
        ], 
        "vpn_target_type": [], 
        "vpn_target_value": [], 
        "vrf": "vpna", 
        "vrf_aftype": [
            "ipv4uni"
        ]
    }, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": false, 
            "evpn": "false", 
            "host": "ce_6850", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "provider": null, 
            "route_distinguisher": null, 
            "ssh_keyfile": null, 
            "state": "absent", 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "username": "rootDC", 
            "validate_certs": true, 
            "vpn_target_state": null, 
            "vpn_target_type": null, 
            "vpn_target_value": null, 
            "vrf": "vpna", 
            "vrf_aftype": "v4"
        }, 
        "module_name": "ce_vrf_af"
    }, 
    "proposed": {
        "route_distinguisher": [], 
        "state": "absent", 
        "vpn_target_type": [], 
        "vpn_target_value": [], 
        "vrf": "vpna", 
        "vrf_aftype": []
    }, 
    "updates": [
        "ip vpn-instance vpna", 
        "undo ipv4-family"
    ]
}
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_vrf_af.py [hunk 1]
```diff
@@ -0,0 +1,856 @@
...
+---
+module: ce_vrf_af
+version_added: "2.4"
+short_description: Manages VPN instance address family on HUAWEI CloudEngine switches.
+description:
+    - Manages VPN instance address family of HUAWEI CloudEngine switches.
+author: Yang yang (@CloudEngine-Ansible)
+notes:
+    - If no vrf is supplied, the module will return error.
+      If state=absent, the vrf will be removed, regardless of the
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vrf_af.py]: `I(state=absent)`


http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

File: lib/ansible/modules/network/cloudengine/ce_vrf_af.py [hunk 2]
```diff
@@ -0,0 +1,856 @@
...
+DOCUMENTATION = '''
+---
+module: ce_vrf_af
+version_added: "2.4"
+short_description: Manages VPN instance address family on HUAWEI CloudEngine switches.
+description:
+    - Manages VPN instance address family of HUAWEI CloudEngine switches.
+author: Yang yang (@CloudEngine-Ansible)
+notes:
+    - If no vrf is supplied, the module will return error.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vrf_af.py]: Not sure if this line is needed as `vrf` has `required: true`

File: lib/ansible/modules/network/cloudengine/ce_vrf_af.py [hunk 3]
```diff
@@ -0,0 +1,856 @@
...
+                            for rtarget in rtargets:
+                                if rtarget.tag in ["vrfRTValue", "vrfRTType"]:
+                                    rt_dict[rtarget.tag] = rtarget.text
+                            vrf_af_info["evpnTargets"].append(rt_dict)
+                self.vrf_af_info["vpnInstAF"].append(vrf_af_info)
+
+    def check_params(self):
+        """Check all input params"""
+        # vrf check
+        if not self.vrf or not self.vrf_aftype:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vrf_af.py]: If both these options are required then then should be set with `required: true` then Ansible will ensure they are set.

Currently you have `required: false` on `aftype`.

If you change that to `required=True` (and update `DOCUMENTATION`) you can remove this check.

File: lib/ansible/modules/network/cloudengine/ce_vrf_af.py [hunk 4]
```diff
@@ -0,0 +1,856 @@
...
+                                          'i.e. X.X.X.X:number<0-65535> or number<0-65535>:number<0-4294967295>'
+                                          'or number<0-65535>.number<0-65535>:number<0-65535>'
+                                          'or number<65536-4294967295>:number<0-65535>'
+                                          ' but not be 0:0 or 0.0:0.')
+            if not self.vpn_target_state:
+                if self.vpn_target_value or self.vpn_target_type:
+                    self.module.fail_json(
+                        msg='Error: The vpn target state should be exist.')
+            if self.vpn_target_state:
+                if not self.vpn_target_value or not self.vpn_target_type:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vrf_af.py:628]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_together
required_one_of
require_if
```

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_vrf_af.py (hunk 1 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vrf_af.py]: I‘ll fix it. thank you

File: lib/ansible/modules/network/cloudengine/ce_vrf_af.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vrf_af.py]:
```
    vrf_aftype:
        description:
            - VPN instance address family.
        required: false
        choices: ['v4','v6']
        default: v4
```
I(aftype) has default valut.  and the check will remove.

**Review by gundalow — APPROVED** [review]

**Pull request merged**